<a href="https://colab.research.google.com/github/ValentynaGrach/ml-homeworks/blob/main/%D0%94%D0%97__18_Reinforcement_learning_ValentynaGrach.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🔹 **КРОК 1: Встановлення бібліотек**

📌 **Опис бібліотек**

- `gymnasium` — для запуску середовищ RL (включно з LunarLander)  
- `swig` — утиліта, для збірки фізичної бібліотеки `Box2D`  
- `gymnasium[box2d]` — підтримка середовищ LunarLander, BipedalWalker, тощо  
- `stable-baselines3` — для навчання агента PPO, DQN та інших RL-алгоритмів  
- `pyvirtualdisplay`, `opencv-python` — для виводу відео гри в Google Colab без справжнього екрана  
- `xvfb`, `ffmpeg` — віртуальний X-дисплей та інструмент для створення відео

---

In [ ]:
!pip install gymnasium
!pip install swig
!pip install "gymnasium[box2d]"
!pip install stable-baselines3
!pip install pyvirtualdisplay opencv-python
!apt install -y xvfb ffmpeg

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 29.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 374.4/374.4 kB 24.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for box2d-py: filename=box2d_py-2.3.5-cp311-cp311-linux_x86_64.whl size=2379450 sha256=7944db96bc2a118ac5ec0566859032ae59c3783d8e8e2c089c4af8c274300bc6
  Stored in directory: /root/.cache/pip/wheels/ab/f1/0c/d56f4a2bdd12bae0a0693ec33f2f0daadb5eb9753c78fa5308
Successfully built box2d-py
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 184.5/184.5 kB 15.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 111.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 79.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 60.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━

# 🔹 КРОК 2: Імпорт бібліотек

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import gym
import imageio
from pyvirtualdisplay import Display
from base64 import b64encode
from IPython.display import HTML

from stable_baselines3 import PPO
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.callbacks import BaseCallback

# 🔹 КРОК 3: Створення середовища LunarLander-v2

🌍 Створюємо середовище для навчання агента – симулятор посадки ракети.

In [ ]:
env = make_vec_env("LunarLander-v3", n_envs=4)

print("🌍 Середовище LunarLander створено.")

🌍 Середовище LunarLander створено.


# 🔹 КРОК 4: Навчання агента з алгоритмом PPO

🧠 Навчаємо агента з нуля керувати ракетою в середовищі за допомогою PPO.

In [ ]:
model = PPO("MlpPolicy", env, verbose=1)
model.learn(total_timesteps=200_000)
print("Навчання завершено ✅")

Using cuda device


/usr/local/lib/python3.11/dist-packages/stable_baselines3/common/on_policy_algorithm.py:150: UserWarning: You are trying to run PPO on the GPU, but it is primarily intended to run on the CPU when not using a CNN policy (you are using ActorCriticPolicy which should be a MlpPolicy). See https://github.com/DLR-RM/stable-baselines3/issues/1245 for more info. You can pass `device='cpu'` or `export CUDA_VISIBLE_DEVICES=` to force using the CPU.Note: The model will train, but the GPU utilization will be poor and the training might take longer than on CPU.
  warnings.warn(


---------------------------------
| rollout/           |          |
|    ep_len_mean     | 88       |
|    ep_rew_mean     | -171     |
| time/              |          |
|    fps             | 1701     |
|    iterations      | 1        |
|    time_elapsed    | 4        |
|    total_timesteps | 8192     |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 91.9         |
|    ep_rew_mean          | -148         |
| time/                   |              |
|    fps                  | 1026         |
|    iterations           | 2            |
|    time_elapsed         | 15           |
|    total_timesteps      | 16384        |
| train/                  |              |
|    approx_kl            | 0.0070136273 |
|    clip_fraction        | 0.0427       |
|    clip_range           | 0.2          |
|    entropy_loss         | -1.38        |
|    explained_variance   | -0.000654    |
|    learning_r

🔍 **Аналіз навченої моделі**

📉 **Loss**: поступово зменшується — з **342 → 18.6 → 4.2 → 1.16**  
→ модель стабільно навчається, без стрибків  

📈 **ep_rew_mean (середня винагорода)**:
- Початок: **-171**
- Кінець: **+98.6** → 🚀 **значне покращення!**

📏 **ep_len_mean** (середня довжина епізоду):
- Зросла з **88** до **889** → агент живе в рази довше

📊 **explained_variance**:
- З ~**-0.00** до **0.815–0.94** → модель **чудово передбачає цільову функцію**

---

✅ **Висновок**:
🔹 Модель **успішно навчається**  
🔹 Агент **з кожною ітерацією діє краще**  
🔹 Винагорода зросла з негативної до стабільно **позитивної**  
🔹 Навчання пройшло **швидко і ефективно**

---

⏭ **Наступний крок** → виведемо відео, **як агент грає** 🎥

# 🔹 КРОК 5: Візуалізація гри агента (відео)

🎥 Запускаємо один епізод гри, записуємо і показуємо відео.

In [ ]:
# 🧩 Патч для сумісності з NumPy >= 2.0
if not hasattr(np, 'bool8'):
    np.bool8 = np.bool_

display = Display(visible=0, size=(1400, 900))
display.start()

# 🎥 Віртуальний дисплей
env = gym.make("LunarLander-v2", render_mode="rgb_array")
obs = env.reset()
video = []

# 🎮 Створення середовища LunarLander
for _ in range(500):
    action, _ = model.predict(obs, deterministic=True)
    obs, reward, done, info = env.step(action)
    frame = env.render()
    video.append(frame)
    if done:
        break

env.close()

# 🧼 Обробка кадрів
clean_video = []
for frame in video:
    if isinstance(frame, (list, tuple)):
        frame = frame[0]
    if len(frame.shape) == 2:
        frame = np.stack([frame] * 3, axis=-1)
    elif frame.shape[-1] not in [1, 3, 4]:
        continue
    clean_video.append(np.array(frame))

# 💾 Зберігаємо відео
video_path = "/content/lunarlander.mp4"
imageio.mimsave(video_path, clean_video, fps=30)

# 👁 Показуємо відео
mp4 = open(video_path, 'rb').read()
data_url = "data:video/mp4;base64," + b64encode(mp4).decode()
HTML('<video width=600 controls autoplay loop><source src="' + data_url + '" type="video/mp4"></video>')

/usr/local/lib/python3.11/dist-packages/gym/core.py:317: DeprecationWarning: WARN: Initializing wrapper in old step API which returns one bool instead of two. It is recommended to set `new_step_api=True` to use new step API. This will be the default behaviour in future.
  deprecation(
/usr/local/lib/python3.11/dist-packages/gym/wrappers/step_api_compatibility.py:39: DeprecationWarning: WARN: Initializing environment in old step API which returns one bool instead of two. It is recommended to set `new_step_api=True` to use new step API. This will be the default behaviour in future.
  deprecation(
